### ISSUE TITLE
Slow stream loading problem

### ISSUE THREAD

**Issue by iongchun** [diaspora/diaspora#7981]
Stream loading on my pod is never fast, but since death of my newer NAS, this problem becomes unbearable now. The database I use is PostgreSQL, so I use slow query log with 5 seconds threshold to detect slow statements.

The first slow query I found is below:
SELECT DISTINCT posts.*, posts.id FROM "posts" INNER JOIN "taggings" ON "taggings"."taggable_id" = "posts"."id" AND "taggings"."taggable_type" = 'Post' WHERE "posts"."type" IN ('StatusMessage') AND "posts"."public" = 't' AND (taggings.tag_id IN (1001,1002,1003,...)) AND (posts.author_id NOT IN (2001,2002,2003,...)) AND (posts.created_at < '2019-02-24 11:07:52') AND "posts"."type" IN ('StatusMessage', 'Reshare') ORDER BY posts.created_at DESC, posts.id DESC LIMIT 15

Query plan:
 Limit  (cost=216691.03..216691.70 rows=15 width=1014)
   ->  Unique  (cost=216691.03..219016.67 rows=51681 width=1014)
         ->  Sort  (cost=216691.03..216820.23 rows=51681 width=1014)
               Sort Key: posts.created_at DESC, posts.id DESC, posts.author_id, posts.public, posts.guid, posts.text, posts.updated_at, posts.provider_display_name, posts.root_guid, posts.likes_count, posts.comments_count, posts.o_embed_cache_id, posts.reshares_count, posts.interacted_at, posts.tweet_id, posts.open_graph_cache_id, posts.tumblr_ids
               ->  Gather  (cost=2618.70..189678.09 rows=51681 width=1014)
                     Workers Planned: 2
                     ->  Nested Loop  (cost=1618.70..183509.99 rows=21534 width=1014)
                           ->  Parallel Bitmap Heap Scan on taggings  (cost=1618.27..58496.78 rows=31524 width=4)
                                 Recheck Cond: (tag_id = ANY ('{1001,1002,1003,...}'::integer[]))
                                 Filter: ((taggable_type)::text = 'Post'::text)
                                 ->  Bitmap Index Scan on index_taggings_on_tag_id  (cost=0.00..1599.35 rows=76204 width=0)
                                       Index Cond: (tag_id = ANY ('{1001,1002,1003,...}'::integer[]))
                           ->  Index Scan using posts_pkey on posts  (cost=0.43..3.97 rows=1 width=1010)
                                 Index Cond: (id = taggings.taggable_id)
                                 Filter: (public AND (created_at < '2019-02-24 11:07:52'::timestamp without time zone) AND ((type)::text = ANY ('{StatusMessage,Reshare}'::text[])) AND ((type)::text = 'StatusMessage'::text) AND (author_id <> ALL ('{2001,2002,2003}'::integer[])))

Notice the unusual sort key on all fields of 'posts' in the plan, and PostgreSQL try to find a unique set, and it hurts performance badly. If DISTINCT is removed, the query becomes:
SELECT  posts.*, posts.id FROM "posts" INNER JOIN "taggings" ON "taggings"."taggable_id" = "posts"."id" AND "taggings"."taggable_type" = 'Post' WHERE "posts"."type" IN ('StatusMessage') AND "posts"."public" = 't' AND (taggings.tag_id IN (1001,1002,1003,...)) AND (posts.author_id NOT IN (2001,2002,2003,...)) AND (posts.created_at < '2019-02-24 11:07:52') AND "posts"."type" IN ('StatusMessage', 'Reshare') ORDER BY posts.created_at DESC, posts.id DESC LIMIT 15

Query plan:
 Limit  (cost=0.86..1336.00 rows=15 width=1014)
   ->  Nested Loop  (cost=0.86..4600087.42 rows=51681 width=1014)
         ->  Index Scan Backward using index_posts_on_created_at_and_id on posts  (cost=0.43..504739.92 rows=1647041 width=1010)
               Index Cond: (created_at < '2019-02-24 11:07:52'::timestamp without time zone)
               Filter: (public AND ((type)::text = ANY ('{StatusMessage,Reshare}'::text[])) AND ((type)::text = 'StatusMessage'::text) AND (author_id <> ALL ('{2001,2002,2003,...}'::integer[])))
         ->  Index Only Scan using index_taggings_uniquely on taggings  (cost=0.43..2.48 rows=1 width=4)
               Index Cond: ((taggable_id = posts.id) AND (taggable_type = 'Post'::text))
               Filter: (tag_id = ANY ('{1001,1002,1003,...}'::integer[]))

Looks like quite an improvement to me. I remove all DISTINCT in the pattern "SELECT DISTINCT table.*", then stream loading becomes so much faster now.
I think the DISTINCT keyword is not necessary in this case, since the primary key is already unique, but maybe I miss something else... ;)

**Comment by goobertron** [comment]
Hi there,
Github is for tracking bugs and other issues in the core software. For support with performance issues, please visit the [podmin support section](https://discourse.diasporafoundation.org/c/support/podmin-support) of our Discourse forum.

**Comment by SuperTux88** [comment]
Removing the `DISTINCT` is a really bad idea, because it's there for a reason and if you remove it, you'll start having smaller or bigger gaps in your stream. Maybe there is another solution to tell postgres not to read all fields?

**Comment by iongchun** [comment]
I think using DISTINCT on several fields, along with an index, is useful in many use cases. But using it on all fields of a table, especially when the primary key is auto-increment...hmm...for example, consider two rows:
Columns: PK,field1,field2,field3
Row#1: 234, a, b, c
Row#2: 678, a, b, c
They are not duplicates of each other because of different PK, even with all other fields equal.

"if you remove it, you'll start having smaller or bigger gaps in your stream."
Why? Shouldn't the result set the same with or without DISTINCT?

**Comment by iongchun** [comment]
I look more into this issue and get more used to the style of ActiveRecord.
I think those DISTINCTs could be replaced by GROUP BY with primary keys, PR here #7984

**Comment by denschub** [comment]
> Why? Shouldn't the result set the same with or without DISTINCT?

Posts can have more than one tag, contacts can be in more than one aspect, ...

**Comment by iongchun** [comment]
> > Why? Shouldn't the result set the same with or without DISTINCT?
> 
> Posts can have more than one tag, contacts can be in more than one aspect, ...

Yes, the case I look into is to retrieve unique posts/person with specified tags/aspects,
so I end up using GROUP BY posts/person.id.

**Comment by jimis** [comment]
My stream is loading forever, so I guess this is still an important usability issue today.

**Comment by Flaburgan** [comment]
It is, indeed. If you want to know more and join the effort please [jump on discourse](https://discourse.diasporafoundation.org/t/stream-performance/2282)
